pkg/virt-handler/hotplug-disk/mount.go

```diff
@@ -458,86 +454,97 @@ func (m *volumeMounter) getBlockFileMajorMinor(fileName string) (int64, int64, s
+func (m *volumeMounter) removeBlockMajorMinor(major, minor int64, manager cgroup.Manager) error {
+	return m.updateBlockMajorMinor(major, minor, false, manager)
 }
 
-func (m *volumeMounter) allowBlockMajorMinor(major, minor int64, path string) error {
-	return m.updateBlockMajorMinor(major, minor, path, true)
+func (m *volumeMounter) allowBlockMajorMinor(major, minor int64, manager cgroup.Manager) error {
+	return m.updateBlockMajorMinor(major, minor, true, manager)
 }
 
-func (m *volumeMounter) updateBlockMajorMinor(major, minor int64, path string, allow bool) error {
-	deviceRule := &configs.DeviceRule{
-		Type:        configs.BlockDevice,
+func (m *volumeMounter) updateBlockMajorMinor(major, minor int64, allow bool, manager cgroup.Manager) error {
+	var err error
+	deviceRule := &devices.Rule{
+		Type:        devices.BlockDevice,
 		Major:       major,
 		Minor:       minor,
 		Permissions: "rwm",
 		Allow:       allow,
 	}
-	if err := m.updateDevicesList(path, deviceRule); err != nil {
-		return err
-	}
-	return nil
-}
-
-func (m *volumeMounter) loadEmulator(path string) (*devices.Emulator, error) {
-	list, err := fscommon.ReadFile(path, "devices.list")
-	if err != nil {
-		return nil, err
-	}
-	return devices.EmulatorFromList(bytes.NewBufferString(list))
-}
 
-func (m *volumeMounter) updateDevicesList(path string, rule *configs.DeviceRule) error {
-	// Create the target emulator for comparison later.
-	target, err := m.loadEmulator(path)
-	if err != nil {
-		return err
-	}
-	target.Apply(*rule)
-
-	file := "devices.deny"
-	if rule.Allow {
-		file = "devices.allow"
-	}
-	if err := fscommon.WriteFile(path, file, rule.CgroupString()); err != nil {
-		return err
-	}
+	rules := append(generateDefaultDeviceRules(), deviceRule)
+	err = manager.Set(&configs.Resources{
+		Devices: rules,
+	})
 
-	// Final safety check -- ensure that the resulting state is what was
-	// requested. This is only really correct for white-lists, but for
-	// black-lists we can at least check that the cgroup is in the right mode.
-	currentAfter, err := m.loadEmulator(path)
 	if err != nil {
-		return err
-	}
-	if !m.skipSafetyCheck {
-		if !target.IsBlacklist() && !reflect.DeepEqual(currentAfter, target) {
-			return errors.New("resulting devices cgroup doesn't precisely match target")
-		} else if target.IsBlacklist() != currentAfter.IsBlacklist() {
-			return errors.New("resulting devices cgroup doesn't match target mode")
-		}
-	}
-	return nil
+		log.Log.Infof("cgroup %s had failed to set device rule. error: %v. rule: %+v", manager.GetCgroupVersion(), err, *deviceRule)
+	} else {
+		log.Log.Infof("cgroup %s device rule is set successfully. rule: %+v", manager.GetCgroupVersion(), *deviceRule)
+	}
+
+	return err
+}
+
+func generateDefaultDeviceRules() []*devices.Rule {
+	if len(defaultDeviceRules) > 0 {
+		// To avoid re-computing default device rules
+		return defaultDeviceRules
+	}
+
+	const permissions = "rwm"
```

Comment:
I see this is the code responsible of permissions there:
```
permissions := "rw"
	if cgroups.IsCgroup2UnifiedMode() {
		// Need 'rwm' permissions otherwise ebpf filtering program attached by runc
		// will deny probing the device file with 'access' syscall. That in turn
		// will lead to libvirtd failure on VM startup.
		// This has been fixed upstream:
		//   https://github.com/opencontainers/runc/pull/2796
		// but the workaround is still needed to support previous versions without
		// the patch.
		permissions = "rwm"
	}
```

I've done the same for cgroups now.